### PR TITLE
[NPM] Decoupling resource cache maps from NSmap

### PR DESCRIPTION
**Pull request by vakalapa**
**Reason for Change**:

**Issue Fixed**:

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests

**Notes**:

### CODE REVIEW

File: npm/nwpolicy.go [hunk 1]
```diff
@@ -125,7 +170,7 @@
 	iptMgr := allNs.iptMgr
 	for _, iptEntry := range iptEntries {
 		if err = iptMgr.Add(iptEntry); err != nil {
-			log.Errorf("Error: failed to apply iptables rule. Rule: %+v", iptEntry)
+			metrics.SendErrorLogAndMetric(util.NetpolID, "[AddNetworkPolicy] Error: failed to apply iptables rule. Rule: %+v", iptEntry)
```
Comment (matmerr) [inline npm/nwpolicy.go:173]: bubble up err

File: npm/nwpolicy.go [hunk 2]
```diff
@@ -77,27 +122,27 @@
...
 		npMgr.DeleteNetworkPolicy(oldPolicy)
 		npMgr.isSafeToCleanUpAzureNpmChain = true
 	}
 
 	// Add (merge) the new policy with others who apply to the same pods
-	if oldPolicy, oldPolicyExists := ns.ProcessedNpMap[hashedSelector]; oldPolicyExists {
+	if oldPolicy, oldPolicyExists := npMgr.ProcessedNpMap[npProcessedKey]; oldPolicyExists {
 		addedPolicy, err = addPolicy(oldPolicy, npObj)
 		if err != nil {
 			log.Logf("Error adding policy %s to %s", npName, oldPolicy.ObjectMeta.Name)
```
Comment (matmerr) [inline npm/nwpolicy.go:135]: metrics

Comment (matmerr) [reply npm/nwpolicy.go:135]: + log err from this conditional

File: npm/nwpolicy.go [hunk 3]
```diff
@@ -148,55 +193,47 @@
...
-		npMgr.NsMap[npNs] = ns
-	}
-
 	_, _, _, ingressIPCidrs, egressIPCidrs, iptEntries := translatePolicy(npObj)
 
 	iptMgr := allNs.iptMgr
 	for _, iptEntry := range iptEntries {
 		if err = iptMgr.Delete(iptEntry); err != nil {
-			log.Errorf("Error: failed to apply iptables rule. Rule: %+v", iptEntry)
+			metrics.SendErrorLogAndMetric(util.NetpolID, "[DeleteNetworkPolicy] Error: failed to apply iptables rule. Rule: %+v", iptEntry)
```
Comment (matmerr) [inline npm/nwpolicy.go:211]: wrap err

File: npm/nwpolicy.go [hunk 4]
```diff
@@ -148,55 +193,47 @@
...
-			ns.ProcessedNpMap[hashedSelector] = deductedPolicy
+			npMgr.ProcessedNpMap[npProcessedKey] = deductedPolicy
 		}
 	}
 
 	if npMgr.canCleanUpNpmChains() {
 		npMgr.isAzureNpmChainCreated = false
 		if err = iptMgr.UninitNpmChains(); err != nil {
-			log.Errorf("Error: failed to uninitialize azure-npm chains.")
+			metrics.SendErrorLogAndMetric(util.NetpolID, "[DeleteNetworkPolicy] Error: failed to uninitialize azure-npm chains with err: %s.", err)
```
Comment (matmerr) [inline npm/nwpolicy.go:236]: remove '.' from log line while we're here

File: npm/nwpolicy_test.go [hunk 5]
```diff
@@ -368,3 +377,34 @@ func TestDeleteNetworkPolicy(t *testing.T) {
...
+					}},
+				},
+			},
+		},
+	}
+
+	netpolKey := GetNetworkPolicyKey(npObj)
+
+	// 2 characters are /
+	if len(netpolKey) <= 2 {
```
Comment (matmerr) [inline npm/nwpolicy_test.go:402]: probs return empty string instead of / if both are empty

File: npm/nwpolicy.go [hunk 6]
```diff
@@ -13,62 +14,118 @@
...
+		err            error
+		ns             *Namespace
+		exists         bool
+		npNs           = util.GetNSNameWithPrefix(npObj.ObjectMeta.Namespace)
+		npName         = npObj.ObjectMeta.Name
+		allNs          = npMgr.NsMap[util.KubeAllNamespacesFlag]
+		timer          = metrics.StartNewTimer()
+		hashedSelector = HashSelector(&npObj.Spec.PodSelector)
+		npKey          = GetNetworkPolicyKey(npObj)
+		npProcessedKey = GetProcessedNPKey(npObj, hashedSelector)
```
Comment (neaggarwMS) [inline npm/nwpolicy.go:91]: What is the difference between processed NP key and raw NP policy, why cant they be same?

Comment (vakalapa) [reply npm/nwpolicy.go:91]: processed Key is based on the pod selector, if different netpols with different names have same pod selector, we are creating a single processedNp aggregating all the rules for the podselector on which we are applying

**Review by neaggarwMS — APPROVED** [review]
:shipit:

**Pull request merged**